java/pattern/constraint/ThingConstraint.java

```diff
@@ -486,10 +486,12 @@ public int hashCode() {
         private final int hash;
 
         Value(TypeQLToken.Predicate predicate, T value) {
-            assert !predicate.isEquality() || value instanceof Comparable || value instanceof ThingVariable<?>;
-            assert !predicate.isSubString() || value instanceof java.lang.String;
             if (predicate == null) throw TypeQLException.of(MISSING_CONSTRAINT_PREDICATE);
             else if (value == null) throw TypeQLException.of(MISSING_CONSTRAINT_VALUE);
+
+            assert !predicate.isEquality() || value instanceof Comparable || value instanceof ThingVariable<?>;
+            assert !predicate.isSubString() || value instanceof java.lang.String;
+
```

Comment:
The assertions before the null checks guarantee that the checks will never succeed. Calling the methods after making sure the variables are non-null seems more sensible.